File: algorithms/cpp/113.cpp
#include <iostream>
#include <vector>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution
{
private:
    void dfs(vector< vector<int> > &result, vector<int> &temp, TreeNode *root, int sum)
    {
        sum -= root->val;
        temp.push_back(root->val);
        if ( root->left == NULL && root->right == NULL && sum == 0 )
            result.push_back(temp);
        if ( root->left )
            dfs(result, temp, root->left, sum);
        if ( root->right )
            dfs(result, temp, root->right, sum);
        temp.pop_back();
        return;
    }
public:
    vector< vector<int> > pathSum(TreeNode* root, int sum)
    {
        vector< vector<int> > result;
        vector<int> temp;
        if ( root )
            dfs(result, temp, root, sum);
        return result;
    }
};

int main()
{
    return 0;
}

File: algorithms/python/113.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def pathSum(self, root, sum):
        """
        :type root: TreeNode
        :type sum: int
        :rtype: List[List[int]]
        """
        if not root: return []
        rls, cur = [], [0]
        def helper(node):
            cur.append(node.val)
            cur[0] += node.val
            if node.left or node.right:
                if node.left: helper(node.left)
                if node.right: helper(node.right)
            elif cur[0] == sum:
                rls.append(cur[1:])
            cur.pop()
            cur[0] -= node.val
        helper(root)
        return rls
